src/sonic-config-engine/minigraph.py

```diff
@@ -98,7 +99,92 @@ def parse_device(device):
+        lcm_comp = lcm_list[0]
+        for i in lcm_list[1:]:
+            lcm_comp = lcm_comp * i / calculate_gcd(lcm_comp, i)
+        lcm_array.append(lcm_comp)
+
+    LCM = sum(lcm_array)
+    return int(LCM)
+
+def calculate_gcd(x, y):
+    while y != 0:
+        (x, y) = (y, x % y)
+    return int(x)
+
+def formulate_ecmp_entry(dpg_ecmp_content, port_device_map):
+    png_ecmp_content = {}
+    FG_NHG_MEMBER = [{}, {}]
+    FG_NHG_PREFIX = [{}, {}]
+    FG_NHG = [{}, {}]
+    FG_NEIGH = [{}, {}]
+    COMP_FG_NHG_MEMBER = {}
+    COMP_FG_NHG_PREFIX = {}
+    COMP_FG_NHG = {}
+    COMP_FG_NEIGH = {}
+    neigh_key_ipv4 = []
+    neigh_key_ipv6 = []
+    ipv4_tag = "fgnhg_v4"
+    ipv6_tag = "fgnhg_v6"
+    port_nhipv4_map = dpg_ecmp_content['port_nhipv4_map']
+    port_nhipv6_map = dpg_ecmp_content['port_nhipv6_map']
+    nhgaddr = dpg_ecmp_content['nhgaddr']
+    nhg_int = dpg_ecmp_content['nhgint']
+    nhipv4_device_map = {port_nhipv4_map[x]: port_device_map[x] for x in port_device_map
+                         if x in port_nhipv4_map}
+    nhipv6_device_map = {port_nhipv6_map[x]: port_device_map[x] for x in port_device_map
+                         if x in port_nhipv6_map}
+    nhipv4_devices = sorted(list(set(nhipv4_device_map.values())))
+    nhipv6_devices = sorted(list(set(nhipv6_device_map.values())))
+
+    nhdevices_ipv4_bank_map = {device: bank for bank, device in enumerate(nhipv4_devices)}
+    nhdevices_ipv6_bank_map = {device: bank for bank, device in enumerate(nhipv6_devices)}
+
+    nhipv4_bank_map = {ip: nhdevices_ipv4_bank_map[device] for ip, device in nhipv4_device_map.items()}
+    nhipv6_bank_map = {ip: nhdevices_ipv6_bank_map[device] for ip, device in nhipv6_device_map.items()}
+
+    ipv4_LCM = calculate_lcm_for_ecmp(nhdevices_ipv4_bank_map, nhipv4_bank_map)
+    ipv6_LCM = calculate_lcm_for_ecmp(nhdevices_ipv6_bank_map, nhipv6_bank_map)
+
+    FG_NHG_MEMBER[0] = {ip: {"FG_NHG": ipv4_tag, "bank": bank} for ip, bank in nhipv4_bank_map.items()}
+    FG_NHG_PREFIX[0] = {nhgaddr[0]: {"FG_NHG": ipv4_tag}}
+    FG_NHG[0] = {ipv4_tag: {"bucket_size": ipv4_LCM}}
+    for ip in nhipv4_bank_map:
+        neigh_key_ipv4.append(str(nhg_int + "|" + ip))
+    FG_NEIGH[0] = {neigh_key: {"family": "IPV4"} for neigh_key in neigh_key_ipv4}
+    FG_NHG_MEMBER[1] = {ip: {"FG_NHG": ipv6_tag, "bank": bank} for ip, bank in nhipv6_bank_map.items()}
+    FG_NHG_PREFIX[1] = {nhgaddr[1]: {"FG_NHG": ipv6_tag}}
+    FG_NHG[1] = {ipv6_tag: {"bucket_size": ipv6_LCM}}
+    for ip in nhipv6_bank_map:
+        neigh_key_ipv6.append(str(nhg_int + "|" + ip))
+    FG_NEIGH[1] = {neigh_key: {"family": "IPV6"} for neigh_key in neigh_key_ipv6}
+    COMP_FG_NHG_MEMBER.update(FG_NHG_MEMBER[0])
+    COMP_FG_NHG_MEMBER.update(FG_NHG_MEMBER[1])
+    COMP_FG_NHG_PREFIX.update(FG_NHG_PREFIX[0])
+    COMP_FG_NHG_PREFIX.update(FG_NHG_PREFIX[1])
+    COMP_FG_NHG.update(FG_NHG[0])
+    COMP_FG_NHG.update(FG_NHG[1])
+    COMP_FG_NEIGH.update(FG_NEIGH[0])
```

Comment:
can we use one set of dictionary here ? can we collapse COMP_FG_NHG and FG_NHG into one ?
similar comment for other also